### PR TITLE
Dtls cleanup

### PR DESCRIPTION
**Pull request by boaks**
Some cleanup and simplification.
The adapted handling of the connection is part of preparing for CID, when the connection must be looked up also with a CID and not only the address:port.  

Signed-off-by: Achim Kraus <achim.kraus@bosch-si.com>

### CODE REVIEW

**Comment by boaks** [comment]
I work on #649 and #644 top on this changes. It would make it less work prone, if this could be merged, because that fix crosses "some" changes here.

File: scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java [hunk 1]
```diff
@@ -779,27 +779,15 @@ private void receiveNextDatagramFromNetwork(DatagramPacket packet) throws IOExce
 		for (final Record record : records) {
 			try {
 
-				switch (record.getType()) {
-				case HANDSHAKE:
-				case APPLICATION_DATA:
-				case ALERT:
-				case CHANGE_CIPHER_SPEC:
-
```
Comment (sbernard31) [inline scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java]: Maybe it was not so bad to remove discard record before to create a new "task" ?

File: scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java [hunk 2]
```diff
@@ -1078,183 +1085,134 @@ private void processChangeCipherSpecRecord(Record record) {
...
-	 * @param record
-	 * @param connection
+	 * @param message handshake message
+	 * @param record record of handshake message
+	 * @param connection connection to process handshake message. May only be
+	 *            {@code null}, if a CLIENT_HELLO is received.
 	 * @throws HandshakeException if the handshake message cannot be processed
 	 */
-	private static void processOngoingHandshakeMessage(final HandshakeMessage message, final Record record, final Connection connection) throws HandshakeException {
+	private void processOngoingHandshakeMessage(final HandshakeMessage message, final Record record, final Connection connection) throws HandshakeException {
```
Comment (sbernard31) [inline scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java]: javadoc say connection can be NULL, if this is true => NPE spotted

File: scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java (hunk 1 above)
Comment (boaks) [reply scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java]: The bad's are already removed in  `List<Record> Record.fromByteArray` by

```
ContentType contentType = ContentType.getTypeByValue(type);
if (contentType == null) {
	LOGGER.debug("Received DTLS record of unsupported type [{}]. Discarding ...", type);
} else {
	records.add(new Record(contentType, version, epoch, sequenceNumber, fragmentBytes, peerAddress));
}
```

File: scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java (hunk 2 above)
Comment (boaks) [reply scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java]: No, javadoc says:

> May only be {@code null}, if a CLIENT_HELLO is received.

The null connection is catched up in `receiveNextDatagramFromNetwork` with

```
if (connection == null && (records.size() != 1 || !records.get(0).isNewClientHello())) {
	LOGGER.debug("Discarding {} records from [{}] received without existing connection", records.size(), peerAddress);
	return;
}
```

(And I plan, to extract the processing of the "connection less CLIENT_HELLO" from the serialized execution, but that will be a future step ... )

**Comment by boaks** [comment]
> (I have not so much time to look at californium, sry for delay)

That's fully OK, you do a very great job for leshan. And if things need to be adjusted later, I do not feel, that this will get a too big issue.

File: scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java (hunk 2 above)
Comment (sbernard31) [reply scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java]: > No, javadoc says:
>
>    May only be {@code null}, if a CLIENT_HELLO is received.

I still understand that when I use this method I can use a null connection.

Comment (boaks) [reply scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java]: And so you propose?
And again, only for a short term, then the processing of "connection less CLIENT_HELLO" will be redesigned in one of the next steps. Then the "maybe null" gets removed.

Comment (sbernard31) [reply scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java]: If this will be removed, ignore this comment. Go to merge :D

Comment (boaks) [reply scandium-core/src/main/java/org/eclipse/californium/scandium/DTLSConnector.java]: > May only be {@code null}

was used in some places, therefore I checked the wrong ones. At this location you are right and I fixed the javadoc. :-)

**Pull request merged**